draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1991,36 +2012,201 @@ conditions.
 
 # Suites for Hashing {#suites}
 
-The following table lists recommended algorithms for different curves and
-encodings. To select a suitable algorithm, choose the encoding associated with
-the target curve. For example, Elligator 2 is the recommended encoding for
-Curve25519, whereas simplified SWU is the recommended encoding for P-256.
-When the hash function is to be used in a protocol whose security is proven in the
-random oracle model, applications SHOULD use the Random Oracle construction
-given in {{roadmap}}.
-
-A suite is a bundle of algorithms that provides concrete recommendations for
-hashing bit strings into points of specific elliptic curve groups. Each suite is
-a tuple (E, H, f, ROM) such that
-
--   E, is the elliptic curve group.
--   H, is the cryptographic hash function used by hash\_to\_base.
--   f, is an encoding function compatible with E.
--   ROM, is a boolean flag indicating whether or not to use the random oracle construction.
-
-This document describes the following set of ciphersuites
-
-| Suite ID | E | H | f | ROM |
-|----------|---|---|---|-----|
-| H2C-0001 | P256         | SHA256 | Simplified SWU           | True |
-| H2C-0002 | P384         | SHA512 | Icart                    | True |
-| H2C-0003 | curve25519   | SHA512 | Elligator 2              | True |
-| H2C-0004 | curve448     | SHA512 | Elligator 2              | True |
-| H2C-0005 | edwards25519 | SHA512 | Elligator 2              | True |
-| H2C-0006 | edwards448   | SHA512 | Elligator 2              | True |
-| H2C-0007 | SECP256K1    | SHA512 | Shallue-van de Woestijne | True |
-| H2C-0008 | BLS12381     | SHA512 | Simplified SWU           | True |
+This section lists recommended suites for hashing to standard elliptic curves.
+
+A suite fully specifies the procedure for hashing bit strings to
+points on a specific elliptic curve group.
+Each suite comprises the following parameters:
+
+- Suite ID, a short name used to refer to a given suite.
+- E, the target elliptic curve over a field F.
+- p, the characteristic of the field F.
+- m, the extension degree of the field F.
+- H, the hash function used by hash\_to\_base.
+- W, the number of evaluations of H in hash\_to\_base.
+- f, a mapping function given in {{mappings}}.
+- h\_eff, the value by which the output of f is multiplied
```

Comment:
what is the meaning of the `eff` part?
`h` is simpler.
Even tough, the distinction between `h` and `h_eff` only differs for pairing curves on the specific case of G2.

Reply:
"eff" is for "effective," because in some cases h_eff != h. I'm trying to avoid overloading h, which we use elsewhere to mean the integer such that n = h * r.

Do you think there's a better alternative that doesn't overlap with h? maybe `h'`?

Reply:
For any suite for which `h=h_eff`, list only `h`.
For the special cases, you can list both constants: the cofactor, (e.g. `h=4`) and the effective cofactor (e.g. `h_eff = h * 10 = 40`)

Reply:
I think I didn't make my rationale clear (which probably means the text needs to be improved!):

- `h` is *not* an explicit part of the description of a suite (though it's implicit in E and F, of course). This is intentional: it would be redundant to include it, and it would cause confusion in the case that `h_eff` != `h`.

- In contrast, `h_eff` is not (always) implicitly defined by the other suite parameters, because some curves admit faster cofactor clearing methods.

- The purpose of specifying `h_eff` is to ensure that all users of a suite clear the cofactor in a compatible way without demanding that everyone use the most optimal implementation. The reason to specify the cofactor clearing method as a scalar is that this is the "least common denominator": it seems reasonable to assume that anyone who can hash to the curve can also do scalar multiplication.

So: I think it would be both unnecessary and confusing to include `h` in any suite, and as a result I really don't like the idea of specifying `h_eff` in terms of `h`.

I will make some edits, both here and potentially in {{cofactor-clearing}}, to make all of the above clearer.

Reply:
I pushed some edits that I hope clarifies and addresses this concern.